### ISSUE TITLE
Ansible-playbook with local connection cannot become user for postgresql_user

### ISSUE THREAD

**Issue by pbudzon** [ansible/ansible-modules-core#4092]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

postgresql_user
become?
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

Default.
##### OS / ENVIRONMENT

Ubuntu 14.04. Problem does not seem platform specific.
##### SUMMARY

Running `ansible-playbook -c local playbook.yml` with postgresql_user and become fails:

```
fatal: [localhost]: FAILED! => {"failed": true, "msg": "Failed to change ownership of the temporary files Ansible needs to create despite connecting as root.  Unprivileged become user would be unable to read the file."}
```
##### STEPS TO REPRODUCE

The following playbook:

```

---
- name: test
  hosts: localhost
  remote_user: root
  become: yes
  gather_facts: no
  tasks:
  - name: Change main user password
    postgresql_user:
      name: postgres
      password: '123'
    become: true
    become_user: postgres
```

run as `ansible-playbook -c local playbook.yml`

fails with the following error:

```
fatal: [localhost]: FAILED! => {"failed": true, "msg": "Failed to change ownership of the temporary files Ansible needs to create despite connecting as root.  Unprivileged become user would be unable to read the file."}
```

**This playbook worked before, although I'm not certain which ansible version was it, possibly 2.0.1**  

What's interesting running this playbook as root:

```
sudo ansible-playbook -c local playbook.yml 
```

provides a correct result:

```
TASK [Change main user password] ***********************************************
changed: [localhost]
```

What's even more annoying, running this ad-hoc (from NON-ROOT account!):

```
ansible localhost -c local -m postgresql_user -a "name=postgres password=234" --become --become-user=postgres
```

works too!

```
localhost | SUCCESS => {
    "changed": true, 
    "user": "postgres"
}
```
##### EXPECTED RESULTS

The playbook should run without errors.
##### ACTUAL RESULTS

Full output with -vvvv:

```
 [WARNING]: provided hosts list is empty, only localhost is available

Loaded callback default of type stdout, v2.0

PLAYBOOK: playbook.yml *********************************************************
1 plays in playbook.yml

PLAY [test] ********************************************************************

TASK [Change main user password] ***********************************************
task path: /tmp/packer-provisioner-ansible-local/playbook.yml:8
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ubuntu
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /tmp/ansible-tmp-1467406185.5-169366582280521 `" && echo ansible-tmp-1467406185.5-169366582280521="` echo /tmp/ansible-tmp-1467406185.5-169366582280521 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpJ9Ei9W TO /tmp/ansible-tmp-1467406185.5-169366582280521/postgresql_user
<127.0.0.1> EXEC /bin/sh -c 'chown -R postgres /tmp/ansible-tmp-1467406185.5-169366582280521/ && sleep 0'
fatal: [localhost]: FAILED! => {"failed": true, "msg": "Failed to change ownership of the temporary files Ansible needs to create despite connecting as root.  Unprivileged become user would be unable to read the file."}

NO MORE HOSTS LEFT *************************************************************
    to retry, use: --limit @playbook.retry

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   
```

**Comment by abadger** [comment]
The local connection plus remote_user is what's confusing ansible.  Local connection ignores the remote_user but it doesn't clear it so the code that's checking whether you should have permission to chown thinks that you are going to be able to connect as root rather than being connected as a normal user.  When chown fails even though it thinks you are root it gives up, assuming that there must be something really wrong.

Since the unpriivleged user to unprivileged user case works, that probably means that you do have access to POSIX filesystem acls and the setfacl command (which is what Ansible tries to use for an unprivileged user).  We've had some updates to the code between 2.1.0 and devel that appear to fix this: https://github.com/ansible/ansible/commit/8c7ceaab816652cef68a25342d4ba51821d6500c  (I jsut tested with remote_user: root and a local connection before and after applying this patch.  Before it fails with the same error you're getting.  After, it succeeds).

I'll ask @jimi-c and @bcoca about cherry-picking the patch to the stable-2.1 branch.  As a workaround you should be able to apply that patch to your local copy of ansible to verify that it fixes things for you.

**Comment by abadger** [comment]
@jimi-c and I discussed this and we think that the more correct thing to do is to set the remote_user to the user invoking ansible when connection=local.  He's going to look into writing a PR to implement that.

**Comment by jimi-c** [comment]
Fixed by 9248cde2398e3dd604d3d079e8fe0061d72c9a19 (and cherry-picked to stable-2.1, so it will be included in 2.1.1).

f you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
